proposals/0000-existentials.rst

```diff
@@ -0,0 +1,504 @@
+          -- Γ ⊢ expr : inner_ty[witness_ty/bound_tv]
+          -- -----------------------------------------------------------------------
+          -- Γ ⊢ Pack witness_ty expr (bound_tv:ki) inner_ty : exists bound_tv. inner_ty
+
+        | Open (Expr b)
+          -- typing rule:
+          -- Γ ⊢ expr : exists (a :: k). inner_ty
+          -- -------------------------------------------------------------------------
+          -- Γ ⊢ Open expr : inner_ty[Witness @k (exists (a :: k). inner_ty) expr / a]
+
+      These typing rule are ``CE-Pack`` and ``CE-Open`` from Fig. 7 of the paper_.
+
+   #. Constructs in Core that bind terms need to update their typing rules to check for
+      skolem escape. For example, here is the updated rule for lambda::
+
+        Γ ⊢ ty1 : TYPE rep
+        rep is a monomorphic representation
+        Γ, (var:ty1) ⊢ expr : ty2
+        var # ty2            -- this is the new check
+        ------------------------------
+        Γ ⊢ Lam (var:ty1) expr : ty1 -> ty2
+
+      Similar changes will be necessary for case alternatives and for ``let``. See ``CE-Abs``
+      in Fig. 7 of the paper_.
+
+   #. We need to add a new coercion form to sllow for an interpretation for ``Witness``.
+      This would be the new constructor for ``Coercion``::
+
+        | WitnessPackCo Type CoreExpr TyVar Type
+          -- typing rule:
+          -- Γ ⊢ Pack witness_ty expr (bound_tv:ki) inner_ty : exists (bound_tv :: ki). inner_ty
+          -- ------------------------------------------------------------------------------------------------------------------------
+          -- Γ ⊢ WitnessPackCo witness_ty expr (bound_tv:ki) inner_ty : Witness witness_ty expr (bound_tv:ki) inner_ty ~# witness_ty
+
+      See ``CG-ProjPack`` from Fig. 7 of the paper_.
+
+   #. Several new coercion forms are necessary in order to support ``liftCoSubst``.
+      These are all added to ``Coercion``::
+
+        | ExistsCo TyVar Coercion    -- lifts ExistsTy
+        | WitnessCo ExprCoercion     -- lifts WitnessTy
+
+        data ExprCoercion                -- witnesses the equality between two expressions
+          = GReflEC CoreExpr CoercionR
+            -- typing rule:
+            -- Γ ⊢ expr |> co : ty
+            -- ------------------------------------------
+            -- Γ ⊢ GReflEC expr co : expr ~ (expr |> co)
+
+          | TransformEC CoreExpr CoreExpr String
+            -- typing rule:
+            -- Γ ⊢ e1 : ty
+            -- Γ ⊢ e2 : ty
+            -- --------------------------------------
+            -- Γ ⊢ TransformEC e1 e2 reason : e1 ~ e2
+
+      The ``TransformEC`` form allows us to create coercions witnessing the equality of
+      any two expressions. The idea is that we use this when we need to prove that an
+      optimization GHC performs is sound. We put the name of the optimization in the
+      carried string. Perhaps in the future, this will become more principled.
```

Comment:
This is still limiting what GHC can do because for example if we have a transformation that removes an unused argument in a function call (e.g. due to absent usage annotation) and then would float out an expression out of a lambda binding that parameter, then now it still needs to record the old expression in the coercion, so the parameter stays used and floating is limited.

Not a show stopper, just saying that I expect existential to sometimes get in the way of optimizations even with this very general coercion.

Reply:
Yes, I see your point here. I have added a Drawback about this. I think it can be worked around (as I write in the proposal).

Reply:
The workaround is interesting, thanks!

Reply:
Hmm, given the workaround, do we really need this very unsafe coercion? Or could we say that any core2core transformation either needs to be consistent (i.e. translate the same expression the same way in all contexts, including in the types, and this way the typing is preserved), or it need to let-bind these expressions first?

Reply:
The work-around is this, right:

> Because of the ability to substitute in the types of a let expression, the appearance of a locally bound variable in a TransformEC need not stop floating. Just float out and replace the variable with its right-hand side.

For recursive lets that’s possible but tedious – you’d have to substitute something like `let x = … in x`.

But what about floating out of lambdas? There is nothing to substitute there?

Reply:
It would seem in the general case floating things out will result in parameters that must partially coincide?
```
foo :: forall x. foreach (a :: _) (b :: _). (a ~ (...x...), b ~ (...x...)) => ...
```
Wouldn't take a fancy combination of floating and common subexpression elimination avoid this?